src/basicfuns.jl

```diff
@@ -152,9 +152,50 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x ≤ x0
+        return exp(x)
+    elseif x ≤ x1
+        return log1p(exp(x))
+    elseif x ≤ x2
+        return x + exp(-x)
+    else
+        return oftype(x0, x)
+    end
+end
+
+#=
+returns a thresholds x0, x1, x2 such that:
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+=#
+@inline @generated function _log1pexp_thresholds(::T) where {T<:Real}
```

Comment:
> Can you please benchmark the loading and compilation of LogExpFunctions and log1pexp, on master and with this change?

Sure, though I'm not sure how to do that :). How do you benchmark loading / compilation time of a package / function?
OTOH, unfortunately I don't see a nice way to avoid `@generated` here.

> Did you notice that the new thresholds are different for Float32?

They are slightly different. On master they are `(_, 9.0, 16.0)` (the first threshold is not used). After this PR,

```
_log1pexp_thresholds(::Float32) # (-16.635532f0, 7.9711924f0, 13.993f0)
```

> Additionally, I think it is better to help the compiler and hardcode the thresholds for `Float64`, `Float32`, etc. E.g., (not sure if dispatching with generated functions is problematic)

I don't see any issues, as far as I know generated functions dispatch fine. With `@generated` it's close to the hard-coded version:

```
julia> @code_typed LogExpFunctions._log1pexp_thresholds(1.0)
CodeInfo(
1 ─     return (-36.7368005696771, 18.021826694558577, 33.231118823529634)
) => Tuple{Float64, Float64, Float64}
```

Reply:
Though hard-coding thresholds for Float32, Float64 is a way to make absolutely sure this PR is not breaking ...

Reply:
> Sure, though I'm not sure how to do that :). How do you benchmark loading / compilation time of a package / function?

Similar to e.g. https://github.com/JuliaDiff/ChainRulesCore.jl/pull/356#issuecomment-849956079 (and probably best to use `--startup-file=no` when starting Julia).

> I don't see any issues, as far as I know generated functions dispatch fine. With @generated it's close to the hard-coded version:

The issue is that the compiler has to work out the values on it's own - which increases compilation time. If you provide them, that's not needed for these common types.

Reply:
I'm fine with slightly different bounds (would be nice to make a proper analysis for how close the values are in this range but they seem sufficiently close), the main motivation is to avoid having to compile the generated function for these types.

Reply:
Ok I introduced the hard-coded thresholds for Float32, Float64, with a comment explaining why. To be extra safe, I also set them to the values they had before this PR for those two cases.

Reply:
> So it's not clear that the values are only computed once, they might be recomputed the next time the function is called.

In this case you can see from `@code_typed` that the thresholds are cached fine.

Reply:
You can't draw any such conclusions from `@code_typed`. It only shows the result of code generation but does not tell you if it will be recomputed or cached. And even if it is fine for your specific Julia version for Float64, according to the docs you should not rely on that it's cached in the same way for other types and/or Julia versions.